File: analysis/flight_recorder_analysis/detectors/tool_misuse.py
"""
Tool-misuse detector.

Flags agents that exhibit any of the following patterns within a session:

1. **Retry storm** — the same ``(tool_name, canonical_args)`` pair is called
   more than ``max_retries`` times and the individual ``retry_count`` field
   confirms repeated attempts.

2. **High failure rate** — more than ``min_failure_rate`` fraction of all
   calls to a given tool fail (requires at least ``min_calls_for_rate``
   samples).

3. **Invalid argument pattern** — tool_call arguments are empty / None on a
   non-first attempt (suggesting the agent is not adapting its inputs despite
   prior failures).

The detector is intentionally conservative: it only fires when *multiple*
signals align, to keep false-positive rates low.
"""

from __future__ import annotations

import json
from collections import defaultdict

from flight_recorder_analysis.config import ToolMisuseConfig
from flight_recorder_analysis.detectors.base import BaseDetector
from flight_recorder_analysis.models import Detection, DetectionType, Event, Severity


def _canonical_args(args: object) -> str:
    """Stable string representation of tool arguments for grouping."""
    if args is None:
        return ""
    try:
        return json.dumps(args, sort_keys=True, ensure_ascii=False)
    except (TypeError, ValueError):
        return str(args)


class ToolMisuseDetector(BaseDetector):
    """Detects tool misuse: retry storms and persistent high failure rates."""

    def __init__(self, config: ToolMisuseConfig | None = None) -> None:
        self._cfg = config or ToolMisuseConfig()

    # ------------------------------------------------------------------
    # BaseDetector interface
    # ------------------------------------------------------------------

    @property
    def name(self) -> str:
        return "ToolMisuseDetector"

    @property
    def version(self) -> str:
        return "1.0.0"

    @property
    def detection_type(self) -> DetectionType:
        return DetectionType.TOOL_MISUSE

    @property
    def description(self) -> str:
        return (
            "Flags agents that call the same tool repeatedly with failing "
            "arguments, or that exhibit a high per-tool failure rate."
        )

    # ------------------------------------------------------------------
    # Detection logic
    # ------------------------------------------------------------------

    def detect(self, events: list[Event]) -> list[Detection]:
        if not events:
            return []

        tool_events = [e for e in events if e.type == "tool_call" and e.tool_call]
        if not tool_events:
            return []

        detections: list[Detection] = []

        # ------------------------------------------------------------------
        # 1. Per-(tool_name, canonical_args) retry storm analysis
        # ------------------------------------------------------------------
        # key → list of (event_id, success, retry_count)
        call_groups: dict[tuple[str, str], list[tuple[str, bool, int]]] = defaultdict(
            list
        )

        for e in tool_events:
            tc = e.tool_call  # type: ignore[union-attr]
            tool_name: str = tc.get("tool_name", "")
            args_key = _canonical_args(tc.get("arguments"))
            success: bool = bool(tc.get("success", False))
            retry_count: int = int(tc.get("retry_count", 0))
            call_groups[(tool_name, args_key)].append((e.id, success, retry_count))

        for (tool_name, args_key), calls in call_groups.items():
            failures = [(eid, rc) for eid, ok, rc in calls if not ok]
            if len(failures) > self._cfg.max_retries:
                max_rc = max(rc for _, rc in failures)
                confidence = min(
                    0.6 + (len(failures) - self._cfg.max_retries) * 0.05, 1.0
                )
                detections.append(
                    Detection(
                        type=DetectionType.TOOL_MISUSE,
                        severity=Severity.WARNING,
                        confidence=round(confidence, 2),
                        message=(
                            f"Tool '{tool_name}' called with identical arguments "
                            f"{len(failures)} times and failing each time "
                            f"(threshold: {self._cfg.max_retries})"
                        ),
                        event_ids=[eid for eid, _ in failures],
                        evidence=(
                            f"args_key={args_key!r}, "
                            f"failure_count={len(failures)}, "
                            f"max_retry_count_seen={max_rc}"
                        ),
                    )
                )

        # ------------------------------------------------------------------
        # 2. Per-tool failure-rate analysis
        # ------------------------------------------------------------------
        # key → (total_calls, failed_calls, event_ids_of_failures)
        tool_stats: dict[str, list[tuple[str, bool]]] = defaultdict(list)
        for e in tool_events:
            tc = e.tool_call  # type: ignore[union-attr]
            tool_name = tc.get("tool_name", "")
            success = bool(tc.get("success", False))
            tool_stats[tool_name].append((e.id, success))

        for tool_name, calls in tool_stats.items():
            total = len(calls)
            if total < self._cfg.min_calls_for_rate:
                continue
            failed_ids = [eid for eid, ok in calls if not ok]
            failure_rate = len(failed_ids) / total
            if failure_rate >= self._cfg.min_failure_rate:
                # Avoid double-reporting the same tool already caught above.
                already_reported = any(
                    tool_name in d.message and d.type == DetectionType.TOOL_MISUSE
                    for d in detections
                )
                if not already_reported:
                    confidence = min(0.5 + failure_rate * 0.4, 1.0)
                    detections.append(
                        Detection(
                            type=DetectionType.TOOL_MISUSE,
                            severity=Severity.WARNING,
                            confidence=round(confidence, 2),
                            message=(
                                f"Tool '{tool_name}' has a {failure_rate:.0%} failure "
                                f"rate over {total} calls "
                                f"(threshold: {self._cfg.min_failure_rate:.0%})"
                            ),
                            event_ids=failed_ids,
                            evidence=(
                                f"total_calls={total}, "
                                f"failed_calls={len(failed_ids)}, "
                                f"failure_rate={failure_rate:.3f}"
                            ),
                        )
                    )

        return detections

File: internal/analyzer/detectors/toolmisuse.go
package detectors

import (
	"encoding/json"
	"fmt"

	"github.com/agent-transponder/agent-transponder/internal/analyzer"
	"github.com/agent-transponder/agent-transponder/internal/types"
)

// callRecord holds the result of a single tool invocation.
type callRecord struct {
	eventID    string
	success    bool
	retryCount int
}

// toolStats accumulates stats for a single tool name within a session.
type toolStats struct {
	// per (tool_name, canonical_args) retry-storm tracking
	byArgs map[string][]callRecord // key: canonicalArgs(arguments)
	// per tool_name overall call list (for failure-rate check)
	all []callRecord
	// which (tool_name, args_key) pairs have already emitted a retry-storm finding
	emittedRetryStorm map[string]struct{}
	// whether a failure-rate finding has been emitted for this tool
	emittedRateDetection bool
}

func newToolStats() *toolStats {
	return &toolStats{
		byArgs:            make(map[string][]callRecord),
		emittedRetryStorm: make(map[string]struct{}),
	}
}

// misuseSess holds per-session state for the ToolMisuseDetector.
type misuseSess struct {
	tools map[string]*toolStats // key: tool_name
}

func newMisuseSess() *misuseSess {
	return &misuseSess{tools: make(map[string]*toolStats)}
}

// ToolMisuseDetector detects tool misuse via two signals (port of tool_misuse.py):
//
//  1. Retry storm: the same (tool_name, canonical_args) pair fails more than
//     maxConsecutiveFailures times.
//
//  2. High failure rate: more than minFailureRate fraction of all calls to a
//     given tool fail (requires at least minCallsForRate samples).
type ToolMisuseDetector struct {
	maxConsecutiveFailures int
	minFailureRate         float64
	minCallsForRate        int
	sessions               map[string]*misuseSess
}

// NewToolMisuseDetector creates a ToolMisuseDetector from plugin config.
// maxConsecutiveFailures maps to max_retries in the Python reference.
// Uses hardcoded defaults for minFailureRate (0.7) and minCallsForRate (3).
func NewToolMisuseDetector(maxConsecutiveFailures int) *ToolMisuseDetector {
	return NewToolMisuseDetectorFull(maxConsecutiveFailures, 0.7, 3)
}

// NewToolMisuseDetectorFull creates a ToolMisuseDetector with all parameters
// explicitly specified.  Useful in tests.
func NewToolMisuseDetectorFull(maxConsecutiveFailures int, minFailureRate float64, minCallsForRate int) *ToolMisuseDetector {
	if maxConsecutiveFailures <= 0 {
		maxConsecutiveFailures = 5
	}
	if minFailureRate <= 0 {
		minFailureRate = 0.7
	}
	if minCallsForRate <= 0 {
		minCallsForRate = 3
	}
	return &ToolMisuseDetector{
		maxConsecutiveFailures: maxConsecutiveFailures,
		minFailureRate:         minFailureRate,
		minCallsForRate:        minCallsForRate,
		sessions:               make(map[string]*misuseSess),
	}
}

// Name implements analyzer.Detector.
func (d *ToolMisuseDetector) Name() string { return "ToolMisuseDetector" }

// Reset implements analyzer.Detector.
func (d *ToolMisuseDetector) Reset() {
	d.sessions = make(map[string]*misuseSess)
}

// Analyze implements analyzer.Detector.
//
//nolint:gocognit // faithful port of tool_misuse.py; extracting sub-functions would obscure the algorithm
func (d *ToolMisuseDetector) Analyze(event *types.Event) ([]analyzer.Finding, error) {
	if event.Type != types.EventToolCall || event.ToolCall == nil {
		return nil, nil
	}

	sess := d.sessions[event.SessionID]
	if sess == nil {
		sess = newMisuseSess()
		d.sessions[event.SessionID] = sess
	}

	tc := event.ToolCall
	toolName := tc.ToolName
	argsKey := canonicalArgs(tc.Arguments)
	rec := callRecord{
		eventID:    event.ID,
		success:    tc.Success,
		retryCount: tc.RetryCount,
	}

	ts := sess.tools[toolName]
	if ts == nil {
		ts = newToolStats()
		sess.tools[toolName] = ts
	}
	ts.byArgs[argsKey] = append(ts.byArgs[argsKey], rec)
	ts.all = append(ts.all, rec)

	var findings []analyzer.Finding

	// -- 1. Retry-storm check ------------------------------------------------
	argsRecords := ts.byArgs[argsKey]
	if _, alreadyEmitted := ts.emittedRetryStorm[argsKey]; !alreadyEmitted {
		var failures []callRecord
		for _, r := range argsRecords {
			if !r.success {
				failures = append(failures, r)
			}
		}
		if len(failures) > d.maxConsecutiveFailures {
			ts.emittedRetryStorm[argsKey] = struct{}{}

			maxRC := 0
			failIDs := make([]string, len(failures))
			for i, f := range failures {
				failIDs[i] = f.eventID
				if f.retryCount > maxRC {
					maxRC = f.retryCount
				}
			}

			confidence := min(0.6+float64(len(failures)-d.maxConsecutiveFailures)*0.05, 1.0)
			findings = append(findings, analyzer.Finding{
				ID:         fmt.Sprintf("misuse-storm-%s-%s-%d", event.SessionID, toolName, len(ts.emittedRetryStorm)),
				SessionID:  event.SessionID,
				AgentID:    event.AgentID,
				TenantID:   event.TenantID,
				Timestamp:  event.Timestamp,
				Detector:   d.Name(),
				Type:       string(types.DetectionToolMisuse),
				Severity:   string(types.SeverityWarning),
				Confidence: round2(confidence),
				Message: fmt.Sprintf(
					"Tool '%s' called with identical arguments %d times and failing each time (threshold: %d)",
					toolName, len(failures), d.maxConsecutiveFailures,
				),
				EventIDs: failIDs,
				Evidence: fmt.Sprintf(
					"args_key=%q, failure_count=%d, max_retry_count_seen=%d",
					argsKey, len(failures), maxRC,
				),
			})
		}
	}

	// -- 2. Failure-rate check -----------------------------------------------
	if !ts.emittedRateDetection {
		total := len(ts.all)
		if total >= d.minCallsForRate {
			var failedIDs []string
			for _, r := range ts.all {
				if !r.success {
					failedIDs = append(failedIDs, r.eventID)
				}
			}
			failureRate := float64(len(failedIDs)) / float64(total)
			if failureRate >= d.minFailureRate {
				// Avoid double-reporting if the retry-storm already fired for this tool.
				alreadyStorm := len(ts.emittedRetryStorm) > 0
				if !alreadyStorm {
					ts.emittedRateDetection = true

					confidence := min(0.5+failureRate*0.4, 1.0)
					findings = append(findings, analyzer.Finding{
						ID:         fmt.Sprintf("misuse-rate-%s-%s", event.SessionID, toolName),
						SessionID:  event.SessionID,
						AgentID:    event.AgentID,
						TenantID:   event.TenantID,
						Timestamp:  event.Timestamp,
						Detector:   d.Name(),
						Type:       string(types.DetectionToolMisuse),
						Severity:   string(types.SeverityWarning),
						Confidence: round2(confidence),
						Message: fmt.Sprintf(
							"Tool '%s' has a %.0f%% failure rate over %d calls (threshold: %.0f%%)",
							toolName, failureRate*100, total, d.minFailureRate*100,
						),
						EventIDs: failedIDs,
						Evidence: fmt.Sprintf(
							"total_calls=%d, failed_calls=%d, failure_rate=%.3f",
							total, len(failedIDs), failureRate,
						),
					})
				}
			}
		}
	}

	return findings, nil
}

// canonicalArgs produces a stable string key for tool arguments by
// unmarshalling and re-marshalling (encoding/json sorts map keys).
func canonicalArgs(raw json.RawMessage) string {
	if raw == nil {
		return ""
	}
	var obj interface{}
	if err := json.Unmarshal(raw, &obj); err != nil {
		return string(raw)
	}
	result, err := json.Marshal(obj)
	if err != nil {
		return string(raw)
	}
	return string(result)
}
